Payment

Eerang2 opened the pull request.

Eerang2 committed:
결제 페이지 구현

임의로 데이터 노출시키고
아임포트 결제창 노출되는지 확인

결제 클릭시  queryDsl 을 사용하여 하나에 로직으로 유저의 존재 확인과 user테이블과 grade테이블 조인 후 등급에따른 할인가 추출 후 프론트에서 넘어온 할인가와 일치하는지 확인

할인이 되는 가격을 계산하여 js 에서 계산한 최종금액과 일치하는지 확인

각 메소드마다 확인된 유저의 정보와 최종가격을 담는 dto 생성하여 response 로 데이터 던짐

Eerang2 committed:
결제 페이지 구현

임의로 데이터 노출시키고
아임포트 결제창 노출되는지 확인

결제 클릭시  queryDsl 을 사용하여 하나에 로직으로 유저의 존재 확인과 user테이블과 grade테이블 조인 후 등급에따른 할인가 추출 후 프론트에서 넘어온 할인가와 일치하는지 확인

할인이 되는 가격을 계산하여 js 에서 계산한 최종금액과 일치하는지 확인

각 메소드마다 확인된 유저의 정보와 최종가격을 담는 dto 생성하여 response 로 데이터 던짐

Eerang2 committed:
Merge remote-tracking branch 'origin/payment' into payment

Eerang2 committed:
git ignore setting

Eerang2 committed:
git ignore setting

Eerang2 committed:
git ignore

Eerang2 committed:
git ignore

Eerang2 committed:
fix : 결제창 이전에 상품검증단계에서 사용자의 등급을 받을때 String 으로 받아서 검증하는 로직을 Enum 으로 받고 즉시 검증으로 코드 변경

검증완료시 주문 테이블에 pending 상태로 저장하고 결제창 내리면 결제실패와 동시에 delete 하는 로직 구현

Eerang2 committed:
git ignore

Eerang2 committed:
ignore

Eerang2 committed:
ignore

Eerang2 committed:
ignore

Eerang2 committed:
fix : pr 에 대한 피드백 수용
 grade 의 검증을 enum 타입으로 받아서 한번에 검증하고 등급에 대한 금액 일치 여부만 확인

 feat : 결제 후 검증 로직
 결제한 후에 imaport api 받아서 포트사로 상품 정보에 대한 여부 확인 후에 상품 고유 번호를 받아서 주문 테이블에 주문 상태를 approve 로 변경

 결제창 내리거나 결제 실패시 대기상태에 있는 주문을 delete 하고 내리지 못한 주문은 매일 새벽 3시에 delete 메서드 동작하게 스케줄 어노테이션 사용

Eerang2 committed:
Merge branch 'master' into payment

Eerang2 committed:
스케쥴 어노테이션 활성화 시키기위한 세팅 어노테이션

eojin312 commented on src/main/java/shopping_mall/application/service/PaymentService.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    // 매일 새벽 3시에 실행 (cron 표현식: "0 0 3 * * *")
+    @Scheduled(cron = "0 0 3 * * *")
```
이런 배치성 로직은 배치용 서비스로 따로 빠져도될듯요

eojin312 commented on src/main/java/shopping_mall/application/service/PaymentService.java:
```diff
@@ -0,0 +1,106 @@
+        IamportResponse<Payment> response = imapotClient.verifyPayment(impUid);
+
+        // 응답 코드가 "00"이면 결제 검증이 성공한 것
+        if (response.getCode() == 0) {
```
response.isSuccess() 로 dto 내부에 메소드 하나 추가해도될듯요

```java
public Boolean isSuccess() {
    return this.code == 0;
    }
 ```

eojin312 replied:
```java
class Payment {

    Integer code;
    BigDecimal paymentCharge;


public Boolean isSuccess() {
    return this.code == 0;
    }
}
```

eojin312 replied:
정정) Payment 는 라이브러리 내 클래스로 수정 불가능

eojin312 commented on src/main/java/shopping_mall/application/service/PaymentService.java:
```diff
@@ -0,0 +1,106 @@
+        // 응답 코드가 "00"이면 결제 검증이 성공한 것
+        if (response.getCode() == 0) {
+            // 검증 성공 시 주문 상태 업데이트
+            orderRepository.updateStateApprove(response.getResponse().getMerchantUid(), ApprovalStatus.APPROVED);
```
이건 굳이 update 문으로 명시했어야하나싶은게,
jpa 에서 transaction 이 종료될때, 더티체킹을 통해 변경감지가 되어서 
알아서 update 구문이 호출될것같아요.

```java
        // Q 주문 조회
        Order order = orderRepository.findByOrderId("orderId")
                .orElseThrow(() -> new HntRuntimeException(OrderError.NOT_FOUND_ORDER));

        // C: 주문 승인처리
        order.changeMerchantUid(response.getResponse().getMerchantUid());
        order.changeStatus(ApprovalStatus.APPROVE);


        // ----------- OrderEntity -------

        public OrderEntity changeMerchantUid(String merchantUid) {
            this.merchantUid = merchantUid;
            return this;
        }

        public OrderEntity changeStatus(ApprovalStatus approvalStatus) {
            this.status = approvalStatus;
            return this;
        }
```

eojin312 replied:
https://jojoldu.tistory.com/415

eojin312 commented on src/main/java/shopping_mall/application/service/PaymentService.java:
```diff
@@ -0,0 +1,106 @@
+            orderRepository.updateStateApprove(response.getResponse().getMerchantUid(), ApprovalStatus.APPROVED);
+            return response.getResponse().getAmount();
+        } else {
+            return null;
```
return null 보단 의미있는 값을 응답해줌이 좋을듯요

BigDecimal.ZERO 라던지?

eojin312 commented on src/main/java/shopping_mall/domain/product/entity/OrderEntity.java:
```diff
@@ -0,0 +1,45 @@
+import java.time.LocalDateTime;
+
+@Entity
+@Table(name = "ORDERS")
```
테이블명은 Orders 말고 Order 가 맞을듯

Eerang2 committed:
추후 장바구니와 주문 테이블의 제작을 고려해 상품과 재고테이블을 분리하여 상품 저장시 재고와 상품이 저장이 되게 로직 수정

Eerang2 committed:
상품 리스트페이지와 상품 디테일 페이지에서 데이터만 노출시키는 로직 구현

Eerang2 committed:
장바구니 기능로직 구현

추후 비회원 기능 구현을 위해 로컬스토리지에 담고
토큰이있을시 장바구니 테이블에 저장함

Eerang2 committed:
장바구니 페이지 구현
로컬 스토리지에 있는 상품키를 받고 상품에 대한 db 정보 조회하고 장바구니res 를 만들어서 띄어줄 데이터 넘김 상품 가격과 수량 계산해서 체크박스에 체크된 상품 가격 실시간 없데이트

Eerang2 committed:
장바구니 결제 페이지 구현

eojin312 commented on src/main/java/shopping_mall/application/payment/impl/SinglePaymentServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    public int deleteByMerchantUid(String merchantUid) {
```
transaction 붙여주세요

eojin312 commented on src/main/java/shopping_mall/application/service/ImapotClient.java:
```diff
@@ -0,0 +1,35 @@
+        Payment payment = response.getResponse();
+
+        if (payment != null) {
+            if (!"paid".equals(payment.getStatus())) {
```
이런 상태값은 enum으로 관리하는게 좋을것같아요
String 으로 상태값을 관리하는건 가독성면에서도 좋지않고, 관리하는면에서도 좋지않아서요

```java
if(PaymentStaus.PAID != payment.getStatus())
//혹은

if(!PaymentStatus.isPaid(payment.getStatus()) {}
```

eojin312 commented on src/main/java/shopping_mall/application/service/ProductService.java:
```diff
@@ -0,0 +1,82 @@
+    public List<ProductWithQuantity> findAllByProducts(List<PaymentReq.CartItem> cartProducts) {
+        List<Long> productIds = cartProducts.stream()
+                .map(PaymentReq.CartItem::getProductId)
+                .collect(Collectors.toList());
```
`findAll()` 에선 `toList()` 를 사용하고
여기선 `.collect(Collectors.toList())` 를 사용했는데,
이유가 있나요 ?

eojin312 commented on src/main/java/shopping_mall/application/service/ProductService.java:
```diff
@@ -0,0 +1,82 @@
+
+        // 상품 ID별 수량 매핑
+        Map<Long, Integer> quantityMap = cartProducts.stream()
+                .collect(Collectors.toMap(
```
```java
Map<Long, Integer> quantityMap = cartProducts.stream()
    .collect(Collectors.toMap(
        PaymentReq.CartItem::getProductId,
        PaymentReq.CartItem::getQuantity,
        Integer::sum // 중복 key 발생 시 수량 합산
    ));
```
